api/osf_groups/serializers.py

```diff
@@ -0,0 +1,192 @@
+
+    def get_unregistered_member(self, obj):
+        unclaimed_records = obj.unclaimed_records.get(self.context['group']._id, None)
+        if unclaimed_records:
+            return unclaimed_records.get('name', None)
+
+    def get_member_method(self, group, role):
+        methods = {
+            MANAGER: group.make_manager,
+            MEMBER: group.make_member,
+        }
+        return methods[role]
+
+    def get_group_role(self, validated_data, default_role):
+        role = validated_data.get('role', default_role)
+        if role not in GROUP_MEMBER_PERMISSIONS:
+            raise exceptions.ValidationError('{} is not a valid role; choose manager or member.'.format(role))
+        return role
+
+    class Meta:
+        type_ = 'group_members'
+
+    def get_absolute_url(self, obj):
+        return absolute_reverse(
+            'osf_groups:group-member-detail',
+            kwargs={
+                'user_id': obj._id,
+                'group_id': self.context['request'].parser_context['kwargs']['group_id'],
+                'version': self.context['request'].parser_context['kwargs']['version'],
+            },
+        )
+
+
+class OSFGroupMemberCreateSerializer(OSFGroupMemberSerializer):
+    id = GroupCompoundIDField(source='_id', required=False, allow_null=True)
+    type = TypeField()
+    full_name = ser.CharField(required=False)
+    email = ser.EmailField(required=False, write_only=True)
+
+    def to_representation(self, instance, envelope='data'):
+        """
+        Use OSFGroupMemberSerializer for the response, but OSFGroupMemberCreateSerializer
+        for the request.  We only want full_name to be writable on create member (for unregistered members).
+        User serializer endpoints should be used to edit user's full_name.
+        """
+        return OSFGroupMemberSerializer(instance=instance, context=self.context).data
+
+    def get_user_object(self, user_id, group):
+        if user_id:
+            user = OSFUser.load(user_id)
+            if not user:
+                raise exceptions.NotFound(detail='User with id {} not found.'.format(user_id))
+            if group.has_permission(user, 'member'):
+                raise exceptions.ValidationError(detail='User is already a member of this group.')
+            return user
+        return user_id
+
+    def create(self, validated_data):
+        group = self.context['group']
+        user = self.get_user_object(validated_data.get('_id', None), group)
+        auth = Auth(self.context['request'].user)
+        full_name = validated_data.get('full_name', None)
+        email = validated_data.get('email', None)
+        role = self.get_group_role(validated_data, MEMBER)
+
+        try:
+            # Making someone a member has the potential to violate the one-manager rule
+            # (This either adds a new user as a member or downgrades a manager to a member)
+            if user:
+                self.get_member_method(group, role)(user, auth)
```

Comment:
Is it possible for an exception to be raised here? Should it be handled?

Reply:
I think everything is handled.  Before it gets here, the group has been verified to be valid, and the role has been enforced to be either `manager` or `member`, which are the keys in the dictionary.   This is also called in a try/except, so if something happens with actually creating the member, the ValueError is caught.